File: StreamingConsumer/src/main/scala/com/epam/sparkconsumer/program/CmdUtils.scala
package com.epam.sparkconsumer.program

import org.apache.commons.cli.{CommandLine, DefaultParser, HelpFormatter, Options}

/**
 * Command line utilities.
 */
object CmdUtils {
  /**
   * Argument constants below.
   */
  val Topic = "topic"
  val Url = "url"
  val FilePath = "filePath"
  val FileFormat = "fileFormat"
  val Help = "help"

  private val HasArg = true
  private val NoArg = false

  /**
   * Possible command line options for this program.
   */
  private val Options = new Options()
    .addOption(Topic, HasArg, "Topic name. Required to run the program.")
    .addOption(Url, HasArg, "Topic url. Required to run the program.")
    .addOption(FilePath, HasArg, "File path. Required to run the program.")
    .addOption(FileFormat, HasArg, "File format. Required to run the program.")
    .addOption(Help, NoArg, "Help. Optional.")

  /**
   * Parses cmd arguments into a CommandLine object.
   *
   * @param args cmd args.
   * @return CommandLine object.
   */
  def parserArgs(args: Array[String]): CommandLine = {
    val cmdParser = new DefaultParser
    cmdParser.parse(Options, args)
  }

  /**
   * Checks if there are enough arguments to run this program.
   *
   * @param cmdLine command line object.
   * @return true or false, telling whether there are enough arguments.
   */
  def areArgumentsGood(cmdLine: CommandLine): Boolean = {
    cmdLine.hasOption(Topic) &&
      cmdLine.hasOption(Url) &&
      cmdLine.hasOption(FilePath) &&
      cmdLine.hasOption(FileFormat)
  }

  /**
   * Prints help.
   */
  def printHelp(): Unit = {
    val helpFormatter = new HelpFormatter
    helpFormatter.printHelp("Consumer.", Options)
  }
}

File: KafkaHomework/src/main/java/com/epam/program/CmdUtils.java
package com.epam.program;

import com.google.common.collect.Lists;
import org.apache.commons.cli.*;

import java.util.List;

/**
 * Utility class for handling cmd args.
 */
public final class CmdUtils {

    /**
     * Private ctor to prevent instantiation.
     */
    private CmdUtils() {

    }

    /**
     * Argument constants below
     */
    private static final boolean HAS_ARG = true;
    private static final boolean NO_ARG = false;

    public static final String TOPIC = "topic";
    public static final String URL = "url";
    public static final String PASS_STORE = "passStore";
    public static final String KEYWORDS = "keywords";
    public static final String HELP = "help";
    public static final String HOW_MANY_MESSAGES = "howManyMessages";

    /**
     * Possible command line options for this program.
     */
    private static final Options OPTIONS = new Options()
            .addOption(TOPIC, HAS_ARG, "Topic name. Required.")
            .addOption(URL, HAS_ARG, "Bootstrap server url. Required.")
            .addOption(PASS_STORE, HAS_ARG, "Password store. Required.")
            .addOption(KEYWORDS, HAS_ARG, "Keywords to search. Required.")
            .addOption(HELP, NO_ARG, "Help. Optional.");

    /**
     * Parses cmd arguments into a command line object.
     *
     * @param args cmd args
     * @return Command line object.
     * @throws ParseException exception if something goes wrong.
     */
    public static CommandLine parseArgs(String[] args) throws ParseException {
        PosixParser parser = new PosixParser();
        return parser.parse(OPTIONS, args);
    }

    /**
     * Extracts keywords to look for in tweets as a List.
     *
     * @param commandLine a CommandLine object.
     * @return keywords to look for in tweets as a List.
     * @throws ParseException in case this arg is not present.
     */
    public static List<String> getKeywords(CommandLine commandLine) throws ParseException {
        String keywordsOneLiner = commandLine.getOptionValue(KEYWORDS);
        if (keywordsOneLiner == null) {
            throw new ParseException("-keywords must be provided.");
        }
        String[] splitWords = keywordsOneLiner.split(",");
        return Lists.newArrayList(splitWords);
    }

    /**
     * Extracts the path to passStore.
     *
     * @param commandLine a CommandLine object.
     * @return passStore path.
     * @throws ParseException in case this arg is not present.
     */
    public static String getPassStore(CommandLine commandLine) throws ParseException {
        String passStore = commandLine.getOptionValue(PASS_STORE);
        if (passStore == null) {
            throw new ParseException("-passStore must be provided.");
        }
        return passStore;
    }

    /**
     * Extracts the topic.
     *
     * @param commandLine a CommandLine object.
     * @return topic name.
     * @throws ParseException in case this arg is not present.
     */
    public static String getTopic(CommandLine commandLine) throws ParseException {
        String topic = commandLine.getOptionValue(TOPIC);
        if (topic == null) {
            throw new ParseException("-topic must be provided");
        }
        return topic;
    }

    /**
     * Extracts bootstrap url.
     *
     * @param commandLine a CommandLine object.
     * @return bootstrap url.
     * @throws ParseException in case this arg is not present.
     */
    public static String getBootstrapUrl(CommandLine commandLine) throws ParseException {
        String bootstrapUrl = commandLine.getOptionValue(URL);
        if (bootstrapUrl == null) {
            throw new ParseException("-url must be provided");
        }
        return bootstrapUrl;
    }

    /**
     * Prints help.
     */
    public static void printHelp() {
        HelpFormatter helpFormatter = new HelpFormatter();
        helpFormatter.printHelp("Twitter kafka client.", OPTIONS);
    }
}
